### ISSUE TITLE
Angular: keen-slider within cdk-virtual-scroll-viewport problem

### ISSUE THREAD

**Issue by lunedam-git** [rcbyr/keen-slider#15]
I realise this is likely not so much an issue as a feature request or cry for help...

My data is a list / array of items, each of which has as one value an array of a varying number of image URLs that I'd like to display in a keen-slider instance. I want to display all my data in a scrollable list.

I created a component that creates an instance of keen-slider from an @Input that is an array of URLs (the images to display). All pretty standard stuff. Works fine.

If I use a standard '*ngFor' style list for my data this works a treat, but my data could grow and cause my list to get very long so I tried to use a 'cdk-virtual-scroll-viewport' and a '*cdkVirtualFor' list within that. This breaks my keen-slider implementation though. It seems to be down to the implementation using '*ngFor' internally to create it's 'slides'. Projecting the 'slides' using ng-content has the same issue (only the final 'slide' is present). I though that maybe using the 'slides' option of keen-slider with a function to create the slide elements might be a way forward, but I can't figure out how to write such a function. From the docs: 

> Function, that return HtmlElement[]

If my 'slides' were a set of fixed elements and didn't have to be created from data then there wouldn't be problem. It just seems that once you try to iterate something to create slides within a component within a '*cdkVirtualFor' list something goes haywire.

**Comment by rcbyr** [comment]
Hello @lunedam-git,

thanks for the explanation of your problem. I'm sorry I don't have an immediate answer. It would be very nice if you could share your code, maybe on CodeSandbox. It would be a lot easier to help you then.

**Comment by lunedam-git** [comment]
Thanks.

I'll see if I can put something together that shows the problem. In the meantime, could you maybe point me at an example of the 'slides' configuration option in use? A function that returns an array of HTMLElemets? Am I reading that right? Have a feeling that if this is the case then it should solve my problem...

**Comment by lunedam-git** [comment]
Hi @rcbyr,

I've put together a Stackblitz demo that I think illustrates the problem I'm having, but in a simpler way than in my actual app. 

The first keen-slider is fine, but the second (within a cdk virtual scroller) somehow has no slides. The third is fine again.

[https://stackblitz.com/edit/angular-ivy-owxio9](url)

**Comment by rcbyr** [comment]
Hey @lunedam-git,

after reading your code example, I'm not sure if I got it right. So I built it the way I thought you'd want to have it. It's quick and dirty but I hope the idea is clear. https://stackblitz.com/edit/angular-ivy-zg2jny

**Comment by lunedam-git** [comment]
Thanks @rcbyr,

Sadly this is an approach I'd already tried and it doesn't work for me. I think I must have simplified my stackblitz a bit too much and it behaved differently to my much larger project. Sorry. Basically the keen-slider is in it's own component (so I don't have to track and dispose of an unknown number of sliders - they can just be created and disposed of within the component). I pass the data (array of URLs) to the component via an @Input.

I'm really keen to see how the 'slides' option is meant to work.

> Function, that return HtmlElement[]

If it can be used as a function to output an array of slide elements, I feel this would perhaps do the trick. Do you have a working example of this? Or have I maybe misunderstood the option?

**Comment by rcbyr** [comment]
I think I will create an example of how to use it. In the default settings, the slider searches for the slides as its children. In some cases, e.g. if the slider markup is created with WebComponents and a ShadowDOM, the slider instance will not find the children. In this case you can pass the slides with a help function.

I think we're getting somewhere. Do you want to build a lazy loading image slider with a large number of images but doesn't want to bloat up the markup with empty imgs/divs?

**Comment by lunedam-git** [comment]
Not exactly, no.

My data is fetched from an external database and basically ends up with an array of objects of x length. x is unknown. It could be an array with 1 object, 15, 50, 500... Each item is an object with several keys. One of the keys has a value of an array of URLs (to remote images). This array can have one URL, 2, 10, ... You get the idea. This array of URLs is what I use for the keen-slider instance for each item in my list.

So my app will look like this:

---------------------
Value from item 1
keen-slider for item 1
---------------------
Value from item 2
keen-slider for item 2
---------------------
Value from item 3
keen-slider for item 3
---------------------
Value from item 4
keen-slider for item 4
---------------------

So basically I end up with a long list, each item of the list having its own keen-slider. The sliders are created inside a component so I don't have to worry about tracking and disposing of them. So the list is created using something like:

<div *ngFor="item of data">
  <p>{{item?.name}}</p>
  <p>{{item?.description}}</p>
  <custom-slider-component [imageURLs]="item?.imageURLs"></custom-slider-component>
</div>

As the number of items fetched from the database could potentially get very large, using a normal '*ngFor' list isn't going to be ideal, but it definitely works.

The solution to the potentially very long list should be to use a 'virtual scroll' (CDK):

<cdk-virtual-scroll-viewport [itemSize]="400">
  <div *cdkVirtualFor="item of data">
  <p>{{item?.name}}</p>
  <p>{{item?.description}}</p>
  <custom-slider-component [imageURLs]="item?.imageURLs"></custom-slider-component>
  </div>
</cdk-virtual-scroll-viewport>

But this does not work and results in:

---------------------
Value from item 1

---------------------
Value from item 2

---------------------
Value from item 3

---------------------
Value from item 4

---------------------

**Comment by rcbyr** [comment]
Ok, I have reread everything you have written and believe that I have now understood the problem. Since I am not familiar with your concrete implementation, I can't give you a definitive solution. But it seems that the slides are not present when the slider is initialized. You can prove this by calling slider.details().size within the "created"-hook. This will probably be zero. To fix this, you need to call the resize() method whenever a new slide is appended. So for your application, you have to find a hook with which you can trigger the resize function. For example, the scrollDispatcher from cdk. 

I will create a new example soon, which will be implemented with the MutationObserver-API. I think this will clarify the usage.

**Comment by lunedam-git** [comment]
Many thanks for your continuing efforts (and patience!).

Sadly this doesn't work. Or is unpredictable. 

I suspect that it's not actually 'keen-slider' but a problem with Angular where the '*ngFor' directive is a problem when inside a component that is part of a '*cdkVirtualFor' created list meaning it takes too long. Which takes me back to the 'slides' option of keen-slider. If this can actually be used to generate an array of 'slide' HTMLElements then this is the trick that will work as I can see and log the array of strings (URLs) passed in to my component that creates the 'keen-slider'. I just need an example of how to do this. What would such a function look like? Or have I misunderstood that option?

To be clear. Everything works fine and as expected all the time if I do not use the CDK 'virtual' scroller. So no <cdk-virtual-scroll-viewport> component and no '*cdkVirtualFor' directive. If I use a plain <div> and an '*ngFor' directive everything is fine. All of my keen-slider instances have all of their slides. Fine for relatively short lists, but when the list gets longer some kind of 'virtual' list is required (or you have to start pagination / infinite lists - which generally looks awful imho).

I'll try to make a better example of what I mean this evening. I see now that Markdown / Github had stripped out quite a lot of the pseudo code I'd typed last night. I've edited it a bit so it might make a little more sense now? I'd really like to see a working implementation if the 'slides' option though. I just can't work out how to programatically create an array of HTMLElements to return from the function.

**Comment by lunedam-git** [comment]
Hi @rcbyr 

This is a lot closer to what I'm trying to do. If you scroll around in the CDK scroller the slider instances start going a bit awry. Likely just me being useless. There are only 10 items in the data here, too.

[https://stackblitz.com/edit/angular-ivy-ccfguu](url)

**Comment by rcbyr** [comment]
Hey @lunedam-git,

I think I made it work. https://stackblitz.com/edit/angular-ivy-qdumgn?file=src%2Fapp%2Fslider%2Fslider.component.ts

**Comment by lunedam-git** [comment]
Hi.

That certainly seems to work on stackblitz which is great. Sadly it doesn't work in my project so clearly there's something else going on and I'm going to have to abandon the whole carousel idea - or more likely switch to a paginated list (yuck!). I can't really carry on wasting your time as there's nothing amiss with keen-slider.

A shame that the 'slides' option of keen-slider obviously can't do what I'd thought it did. I'd still be interested in seeing a working example of what that does though.

Thanks for all your help and time. Keen-slider itself is a great project and I appreciate it.

**Comment by rcbyr** [comment]
Hmm, I'm not quite happy with that. I'm a little busy right now, but I really want to make this work. Even if it's just to learn something. So I will write to you if I find something out.

**Comment by lunedam-git** [comment]
Hey @rcbyr ,

I'm beginning to think that the issue is actually with Firebase? My app gets its' data from Firebase. This is when it goes really wonky. Feed the array of URLs in manually and all is happy. Feed a manually created array into an observable and unwrap them with the async pipe (which is what I do in my app with the observable from Firebase) and all is well. Use the observable actually obtained from Firebase and unwrap that with the async pipe and BOOM! Fine in a 'normal' *ngFor, rubbish in a cdk virtual scroll.

Tried this with a different carousel, too. Same results.

Odd.

**Comment by lunedam-git** [comment]
Hi @rcbyr,

Just an update to let you know that I've solved the issue I was having with the cdk virtual scroll and keen-slider in my app. Not terribly elegant, but it was all down to timing and the fact that the I'm unwrapping observables inside unwrapped observables so I added a 'tick' to the keen-slider creation. Oddly enough I still have to do this twice - still not got the the bottom of that. If I remove either of the creations it all goes wrong again. But looking at the memory etc. it doesn't add any issues or overhead as I'm just redeclaring the same thing I'm happy enough. 

The first keen-slider creation is in ngAfterViewInit and the second is in ngAfterViewChecked. Added the 'tick' by simply wrapping the keen-slider creation inside setTimeOut. My 'virtual' scroll now works as I wanted with each item in the list having a working carousel!

And I upgraded to Angular 10, but I don't think that was a factor.
